### ISSUE TITLE
[Bug]: LoRA linear path (lora_shrink/lora_expand) nondeterministic in serving; only an fp32-eager interior is stable — full elimination matrix

### ISSUE THREAD

**Issue by darth-pika-hu** [vllm-project/vllm#50602]
## Summary

With a single LoRA adapter under greedy decoding, the punica linear path
(`lora_shrink`/`lora_expand`) produces run-to-run different outputs
**only inside the live serving engine**. We ran 15 discriminating
serving arms and 660k+ direct kernel invocations. Every "fast" interior
flips; an fp32-eager-cuBLAS interior with identical math never flips
(64+/64+ byte-identical serving calls across two images and two
dependency stacks). No isolated reproduction exists at any tested shape
or config. We could not identify the final mechanism; we can share the
reproduction matrix.

## Environment

- Model: an NVFP4-quantized (`modelopt_mixed`) MoE vision-language
  model (~35B, hybrid linear-attention/full-attention), single bf16
  LoRA rank 16 (`--max-loras 1`), FP8 KV cache, FlashInfer attention,
  Marlin MoE, MTP speculative decoding
  (`{"method":"mtp","num_speculative_tokens":3,"moe_backend":"triton"}`),
  chunked prefill, temperature 0
- Hardware: GB10 (SM121), aarch64, CUDA 13.0.2
- vLLM v0.26.0 source-built; reproduced identically on
  torch 2.11.0+cu130 / triton 3.6.0 / flashinfer 0.6.14 **and**
  torch 2.13.0+cu130 / triton 3.7.1 / flashinfer 0.6.15.post1
- Probe: one server per arm; 16 byte-identical requests (fixed
  multimodal request, ~9k prompt tokens); byte-compare the 16 outputs.
  Prefix caching OFF (so no cache-state effects).

## Matrix

Interiors that FLIP (2-4 distinct outputs per 16 identical calls,
sporadic):

| Variant | Result |
|---|---|
| stock `lora_shrink`+`lora_expand` (triton) | flips |
| + shrink `split_k=1` (the #50278 change, verified resolving at runtime) | flips |
| + tie-free metadata sort (no equal-key `torch.sort` ties) | flips |
| + `cudagraph_specialize_lora: false` | flips |
| + expand `block_n` 128→64 (cf. #48590) | flips |
| + all LoRAKernelMeta H2D copies `non_blocking=False` | flips |
| + `--async-scheduling` removed | flips |
| custom single-writer triton kernel (fused shrink+expand) | flips |
| fp32 batched cuBLAS (`torch.matmul` broadcast over slices) | flips (1/16) |
| custom single-writer CUDA kernel (fp32 fixed-order, atomics-free) | flips |
| same CUDA kernel fed by `.clone()` snapshots of x/A/B/token-mapping | flips |
| eager bf16 matmuls on live buffers (no dtype casts) | flips |

Interior that NEVER flips (same custom-op boundary, same math):

```python
mask = (tmap.to(torch.int64) >= 0).to(torch.float32)[:x.size(0), None]
t = (x.to(torch.float32) @ lora_a.to(torch.float32).T) * scale
d = (t @ lora_b.to(torch.float32).T) * mask
y[:, off:off+width].add_(d.to(y.dtype))
```

64+/64+ byte-identical serving calls, plus repeated 40-request
benchmark sweeps, across both dependency stacks.

Isolated reproduction attempts (all CLEAN — zero mismatches):

- direct `lora_shrink`/`lora_expand` loops with production configs,
  300k invocations at decode shapes and 60k at prefill shapes (M=8192),
  including single-slice N=4096 at default expand `block_n=128`
- in-process engine probes (offline `LLM()`, both enforce_eager and
  piecewise-cudagraph modes, with and without MTP): bitwise-identical
  logits across repetitions

## Non-causes established

triton version (3.6.0/3.7.1), torch version (2.11/2.13), flashinfer
version, shrink split-K atomics, metadata sort tie order, LoRA cudagraph
specialization, expand tile size, metadata H2D copy asynchrony, async
scheduling, and the kernels themselves in isolation.

## Related

- #50278 proposes shrink `split_k=1`: necessary (it removes a real
  relaxed-atomic reduction) but **not sufficient** — all arms above ran
  with it applied and verified.
- #48862 (LoRA triton + modelopt NVFP4 illegal memory access) and
  #48590 (`lora_expand` block_n=128 NaN on sm_90) may share a root
  cause.

Probe harness details and per-arm logs available on request.
